File: reference/src/hash.ts
// 🔒 決定論の中核: SHA-256・タグ付き派生RNG（docs/01 1.2 / 1.3）。
// バイトオーダは big-endian 固定。タグは UTF-8 NFC 正規化してから連結。
//
// NOTE(移植時の注意): uniform() は u64be/2^64 を double にして返す。JS では決定論的だが、
// 他言語へ移植する際は「同じ substream バイト列から同じ double」になるかを golden vectors で検証すること。
// レア度のしきい値比較などで最終結果が float の最終ビットに依存しないよう、impl-design で
// 整数しきい値化を検討する（docs では未確定の実装論点）。

import { createHash } from "node:crypto";

export function sha256(data: Uint8Array): Buffer {
  return createHash("sha256").update(data).digest();
}

/** tag を UTF-8 NFC 正規化（合成/分解済みの混在を吸収）。 */
function nfc(tag: string): string {
  return tag.normalize("NFC");
}

function u32be(n: number): Buffer {
  const b = Buffer.alloc(4);
  b.writeUInt32BE(n >>> 0, 0);
  return b;
}

/** seed = SHA-256(normalizedKey)（32 bytes）。 */
export function makeSeed(normalizedKey: Uint8Array): Buffer {
  return sha256(normalizedKey);
}

/**
 * 各決定ごとに独立な 256bit を生成。
 * substream = SHA-256( seed || 0x00 || utf8(nfc(tag)) || u32be(counter) )
 */
export function substream(seed: Uint8Array, tag: string, counter = 0): Buffer {
  return sha256(
    Buffer.concat([
      Buffer.from(seed),
      Buffer.from([0x00]),
      Buffer.from(nfc(tag), "utf8"),
      u32be(counter),
    ]),
  );
}

/**
 * 一様実数 [0,1) … 先頭8バイトを big-endian u64 化して 2^64 で割る。
 * 【連続出力専用】（colorMod / preservationScore）。離散判定には使わない（docs/08 8.4）。
 */
export function uniform(seed: Uint8Array, tag: string): number {
  const hi = substream(seed, tag).readBigUInt64BE(0);
  return Number(hi) / 2 ** 64;
}

/**
 * 先頭4バイトを big-endian u32 化（0..2^32-1）。
 * 【離散判定用】整数しきい値との直接比較に使う（docs/08 8.4）。
 */
export function u32(seed: Uint8Array, tag: string, counter = 0): number {
  return substream(seed, tag, counter).readUInt32BE(0);
}

/** 一様整数 [0, n) … 剰余バイアスを避ける棄却法。4バイトを big-endian で抽出。 */
export function uintBelow(seed: Uint8Array, tag: string, n: number): number {
  if (!Number.isInteger(n) || n <= 0) throw new Error(`uintBelow: n must be positive int, got ${n}`);
  const limit = Math.floor(0x1_0000_0000 / n) * n;
  for (let c = 0; ; c++) {
    const x = substream(seed, tag, c).readUInt32BE(0);
    if (x < limit) return x % n;
  }
}

export interface Weighted<T> {
  value: T;
  w: number;
}

/**
 * 重み付き選択（docs/03 3.3 pickWeighted）。【整数版・離散判定】docs/08 8.4。
 * - 重みは非負整数（f64 の uniform*total を使わない）。
 * - 候補は value 昇順に並べてから累積（順序固定＝決定論）。
 * - r = uintBelow(seed, tag, Σw) を累積整数重みで走査。重み0は絶対に選ばれない。
 */
export function pickWeighted<T extends string | number>(
  seed: Uint8Array,
  tag: string,
  items: readonly Weighted<T>[],
): T {
  const sorted = [...items].sort((a, b) =>
    a.value < b.value ? -1 : a.value > b.value ? 1 : 0,
  );
  for (const it of sorted) {
    if (!Number.isInteger(it.w) || it.w < 0)
      throw new Error(`pickWeighted: weights must be non-negative integers (tag=${tag})`);
  }
  const total = sorted.reduce((s, i) => s + i.w, 0);
  if (total <= 0) throw new Error(`pickWeighted: total weight must be > 0 (tag=${tag})`);
  const r = uintBelow(seed, tag, total); // [0, total) 整数
  let acc = 0;
  for (const it of sorted) {
    acc += it.w;
    if (r < acc) return it.value;
  }
  return sorted[sorted.length - 1]!.value; // 到達不能（保険）
}

File: rust/qrac-core/src/hash.rs
//! 🔒 決定論コア: SHA-256・タグ付き派生RNG（docs/01 1.2/1.3）。
//! reference/src/hash.ts と1:1。big-endian 固定・タグNFC・整数しきい値（離散）/ f64（連続）。

use sha2::{Digest, Sha256};
use unicode_normalization::UnicodeNormalization;

pub fn sha256(data: &[u8]) -> [u8; 32] {
    let mut h = Sha256::new();
    h.update(data);
    h.finalize().into()
}

pub fn to_hex(bytes: &[u8]) -> String {
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        s.push_str(&format!("{:02x}", b));
    }
    s
}

/// tag を UTF-8 NFC 正規化（タグは ASCII なので実質 no-op だが規約として固定）。
fn nfc(tag: &str) -> String {
    tag.nfc().collect()
}

pub fn make_seed(normalized_key: &[u8]) -> [u8; 32] {
    sha256(normalized_key)
}

/// substream = SHA-256( seed || 0x00 || utf8(nfc(tag)) || u32be(counter) )
pub fn substream(seed: &[u8], tag: &str, counter: u32) -> [u8; 32] {
    let tag_n = nfc(tag);
    let mut v = Vec::with_capacity(seed.len() + 1 + tag_n.len() + 4);
    v.extend_from_slice(seed);
    v.push(0x00);
    v.extend_from_slice(tag_n.as_bytes());
    v.extend_from_slice(&counter.to_be_bytes());
    sha256(&v)
}

/// 一様実数 [0,1)。【連続出力専用】（colorMod / preservationScore）。
pub fn uniform(seed: &[u8], tag: &str) -> f64 {
    let s = substream(seed, tag, 0);
    let hi = u64::from_be_bytes(s[0..8].try_into().unwrap());
    hi as f64 / 2f64.powi(64)
}

/// 先頭4バイトの big-endian u32。【離散判定用】整数しきい値と直接比較。
pub fn u32v(seed: &[u8], tag: &str, counter: u32) -> u32 {
    let s = substream(seed, tag, counter);
    u32::from_be_bytes(s[0..4].try_into().unwrap())
}

/// 一様整数 [0, n)。剰余バイアス回避の棄却法。big-endian。
pub fn uint_below(seed: &[u8], tag: &str, n: u32) -> u32 {
    assert!(n > 0, "uint_below: n must be > 0");
    let n64 = n as u64;
    let limit = (0x1_0000_0000u64 / n64) * n64;
    let mut c: u32 = 0;
    loop {
        let x = u32v(seed, tag, c) as u64;
        if x < limit {
            return (x % n64) as u32;
        }
        c += 1;
    }
}

/// 重み付き選択（整数版・離散, docs/03 3.3 / docs/08 8.4）。
/// 値昇順に並べてから累積、r = uint_below(total)。重み0は選ばれない。
pub fn pick_weighted(seed: &[u8], tag: &str, items: &[(&str, u32)]) -> String {
    let mut sorted: Vec<(&str, u32)> = items.to_vec();
    sorted.sort_by(|a, b| a.0.as_bytes().cmp(b.0.as_bytes())); // 値の UTF-8 バイト昇順
    let total: u32 = sorted.iter().map(|i| i.1).sum();
    assert!(
        total > 0,
        "pick_weighted: total weight must be > 0 (tag={tag})"
    );
    let r = uint_below(seed, tag, total);
    let mut acc: u32 = 0;
    for (v, w) in &sorted {
        acc += *w;
        if r < acc {
            return v.to_string();
        }
    }
    sorted.last().unwrap().0.to_string()
}
